odl/space/base_tensors.py

```diff
@@ -0,0 +1,741 @@
+# along with ODL.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Base classes for implementations of tensor spaces."""
+
+# Imports for common Python 2/3 codebase
+from __future__ import print_function, division, absolute_import
+from future import standard_library
+standard_library.install_aliases()
+
+import numpy as np
+
+from odl.set.sets import Set, RealNumbers, ComplexNumbers
+from odl.set.space import LinearSpace, LinearSpaceElement
+from odl.util import (
+    is_scalar_dtype, is_floating_dtype, is_real_dtype,
+    is_complex_floating_dtype, safe_int_conv,
+    arraynd_repr, arraynd_str, dtype_str, signature_string, indent_rows)
+from odl.util.ufuncs import TensorSpaceUfuncs
+from odl.util.utility import TYPE_MAP_R2C, TYPE_MAP_C2R
+
+
+__all__ = ('TensorSpace',)
+
+
+class TensorSpace(LinearSpace):
+
+    """Base class for sets of tensors of arbitrary data type.
+
+    A tensor is, in the most general sense, a multi-dimensional array
+    that allows operations per entry (keep the rank constant),
+    reductions / contractions (reduce the rank) and broadcasting
+    (raises the rank).
+    For non-numeric data type like ``object``, the range of valid
+    operations is rather limited since such a set of tensors does not
+    define a vector space.
+    Any numeric data type, on the other hand, is considered valid for
+    a tensor space, although certain operations - like division with
+    integer dtype - are not guaranteed to yield reasonable results.
+
+    Under these restrictions, all basic vector space operations are
+    supported by this class, along with reductions based on arithmetic
+    or comparison, and element-wise mathematical functions ("ufuncs").
+
+    See `this Wikipedia article <https://en.wikipedia.org/wiki/Tensor>`_
+    on tensors for further details.
+    """
+
+    def __init__(self, shape, dtype, order='A'):
+        """Initialize a new instance.
+
+        Parameters
+        ----------
+        shape : positive int or sequence of positive ints
+            Number of entries per axis for elements in this space. A
+            single integer results in a space with rank 1, i.e., 1 axis.
+        dtype :
+            Scalar data type of elements in this space. Can be provided
+            in any way the `numpy.dtype` constructor understands, e.g.
+            as built-in type or as a string. Data types with non-trivial
+            shapes are not allowed.
+        order : {'A', 'C', 'F'}, optional
+            Axis ordering of the data storage. Only relevant for more
+            than 1 axis.
+            For ``'C'`` and ``'F'``, elements are forced to use
+            contiguous memory in the respective ordering.
+            For ``'A'`` ("any") no contiguousness is enforced.
+        """
+        try:
+            self.__shape = tuple(safe_int_conv(s) for s in shape)
+        except TypeError:
+            self.__shape = (safe_int_conv(shape),)
+        if any(s < 0 for s in self.shape):
```

Comment:
Actually, no. Numpy allows size 0, so we should, too. We even have a test that fails after I changed this.